migration/blocktransactions/blocktransactions.go

```diff
@@ -0,0 +1,393 @@
+	)
+	if err != nil {
+		return err
+	}
+
+	fetchedTxCount := len(blockTransactions.Indexes.Transactions)
+	fetchedReceiptCount := len(blockTransactions.Indexes.Receipts)
+
+	if fetchedTxCount == 0 || fetchedReceiptCount == 0 {
+		has, err := core.BlockTransactionsBucket.Has(database, blockNumber)
+		if err != nil {
+			return err
+		}
+		// ALready migrated
+		if has {
+			logger.Infow("skipping already migrated block", "blockNumber", blockNumber)
+			return nil
+		}
+		// Not migrated yet, no transactions found, while there are expected transactions
+		if txCount > 0 {
+			return errors.New("missing transactions and receipts")
+		}
+	}
+
+	if fetchedTxCount != txCount {
+		return fmt.Errorf("invalid transactions: expected %d, got %d", txCount, fetchedTxCount)
+	}
+
+	if fetchedReceiptCount != txCount {
+		return fmt.Errorf("invalid receipts: expected %d, got %d", txCount, fetchedReceiptCount)
+	}
+
+	return core.BlockTransactionsBucket.Put(batch, blockNumber, &blockTransactions)
+}
+
+func deleteOldData(database db.KeyValueRangeDeleter, startBlock, endBlock uint64) error {
+	err := core.TransactionsByBlockNumberAndIndexBucket.Prefix().DeleteRange(
+		database,
+		startBlock,
+		endBlock+1,
+	)
+	if err != nil {
+		return err
+	}
+
+	err = core.ReceiptsByBlockNumberAndIndexBucket.Prefix().DeleteRange(
+		database,
+		startBlock,
+		endBlock+1,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type writeTask struct {
+	batch        db.Batch
+	totalTxCount int
+	err          error
+}
+
+func createBatch(
+	logger utils.SimpleLogger,
+	database db.KeyValueStore,
+	startBlock,
+	endBlock uint64,
+	batch db.Batch,
+) (txCount int, err error) {
+	counters, txCount, err := getTxCount(database, startBlock, endBlock)
+	if err != nil {
+		return 0, err
+	}
+
+	debugLog(logger, "fetching blocks", startBlock, endBlock, txCount)
+	defer debugLog(logger, "fetched blocks", startBlock, endBlock, txCount)
+
+	for blockNumber := startBlock; blockNumber <= endBlock; blockNumber++ {
+		err := readBlock(
+			logger,
+			database,
+			batch,
+			blockNumber,
+			counters[blockNumber-startBlock],
+		)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	if err := deleteOldData(batch, startBlock, endBlock); err != nil {
+		return 0, err
+	}
+
+	return txCount, nil
+}
+
+func debugLog(logger utils.SimpleLogger, msg string, startBlock, endBlock uint64, txCount int) {
+	if startBlock%logRate != 0 {
+		return
+	}
+
+	logger.Debugw(
+		msg,
+		"startBlock", startBlock,
+		"endBlock", endBlock,
+		"txCount", txCount,
+		"numGoroutines", runtime.NumGoroutine(),
+	)
+}
+
+func getFirstBlockInBucket[A any](items iter.Seq2[prefix.Entry[A], error]) (uint64, bool, error) {
+	for item, err := range items {
+		if err != nil {
+			return 0, false, err
+		}
+
+		var blockNumIndex db.BlockNumIndexKey
+		if err := blockNumIndex.UnmarshalBinary(item.Key[1:]); err != nil {
+			return 0, false, err
+		}
+		return blockNumIndex.Number, true, nil
+	}
+	return 0, false, nil
```

Comment:
How about consuming iter manually here to avoid loop, since as far as I understood loop iterates only once, and loop is there for calling `next` once

```suggestion
	next, stop := iter.Pull2(items)
	defer stop()
	item, err, ok := next()
	if !ok {
		return 0, false, nil
	}

	if err != nil {
		return 0, false, err
	}

	var blockNumIndex db.BlockNumIndexKey
	if err := blockNumIndex.UnmarshalBinary(item.Key[1:]); err != nil {
		return 0, false, err
	}
	return blockNumIndex.Number, true, nil
```